### PR TITLE
Replace all uses of wiki.slug in favor of wiki.path, fixes #736

### LINKED ISSUE #736: Replace all uses of wiki.slug in favor of wiki.path
We no longer use `node.slug` and it's redundant with the `node.path` field, so we should remove all instances from the codebase. Here they are:

```ruby
app/views/wiki/show.html.erb:  <div class="alert alert-warning"><%= raw t('wiki.show.revision_from', :time => @revision.created_at.to_s(:long)) %> <a href="/wiki/revisions/<%= @node.slug %>"><%= t('wiki.show.view_all_revisions') %></a></div>
app/views/wiki/show.html.erb:        <li><a href="/talk/<%= @node.slug %>"><i class="fa fa-comments-o"></i><span class="hidden-xs hidden-sm"> <%= t('wiki.show.talk') %></span></a></li>
app/views/wiki/show.html.erb:        <li><a href="/wiki/revisions/<%= @node.slug %>"><span class="hidden-xs"><%= @node.revisions.length %> </span><i class="fa fa-list"></i></a></li>
app/views/wiki/show.html.erb:    <% slug = @node.power_tag('parent') %>
app/views/wiki/show.html.erb:    <% parent = DrupalNode.find_by_slug(slug) %>
app/views/wiki/show.html.erb:          <p><a href="/tag/<%= @node.slug %>"><%= raw t('wiki.show.more_research', :tag => @node.slug) %> &raquo;</a></p>
app/views/wiki/edit.html.erb:      <li><a href="/wiki/<%= wiki.slug %>"><i class="fa fa-book"></i> <%= wiki.title %></a></li>
app/views/wiki/edit.html.erb:      <% if @node.path %><li><a href="/wiki/revisions/<%= @node.slug %>"><%= t('wiki.edit.revisions_for_this_page') %></a></li><% end %>
app/views/wiki/revisions.html.erb:      <li><a href="/wiki/<%= @node.slug %>"><%= t('wiki.revisions.view') %></a></li>
app/views/wiki/revisions.html.erb:      <li><a href="/wiki/edit/<%= @node.slug %>"><%= t('wiki.revisions.edit') %></a></li>
app/views/wiki/revisions.html.erb:      <li class="active"><a href="/wiki/revisions/<%= @node.slug %>">Revisions (<%= @node.revisions.length %>)</a></li>
app/views/talk/show.html.erb:<div><iframe src="https://pad.publiclab.org/p/<%= @node.slug[0,50] %>" style="width: 100%; height:800px;"><p><%= t('talk.show.try_direct_link', :url1 => 'https://pad.publiclab.org/p/'+@node.slug[0,50]) %></p></iframe></div>
app/models/drupal_node.rb:  attr_accessible :title, :uid, :status, :type, :vid, :cached_likes, :comment, :path, :slug
app/models/drupal_node.rb:  friendly_id :friendly_id_string, use: [:slugged, :history]
app/models/drupal_node.rb:    slug.blank? || title_changed?
app/models/drupal_node.rb:    self.slug[0..5] == 'place/'
app/models/drupal_node.rb:  def self.find_root_by_slug(title)
app/controllers/talk_controller.rb:    @node = DrupalNode.find_by_slug params[:id]
app/controllers/wiki_controller.rb:      @wikis = DrupalTag.find_pages(@node.slug,30) if @node.has_tag('chapter') || @node.has_tag('tabbed:wikis')
app/controllers/wiki_controller.rb:    @node = DrupalNode.find_root_by_slug(params[:id])
app/controllers/admin_controller.rb:        redirect_to "/wiki/revisions/<%= @node.slug %>" + '?_=' + Time.now.to_i.to_s
```

Note that each time, we can't just use `path` -- If a wiki has slug `hello-world`, it's supposed to have a path of `/wiki/hello-world`. So we should make a new method on `DrupalNode` like:

```ruby
def slug_from_path
  self.path.split('/').last
end
```

And then replace all of the above instances of `node.slug` with `node.slug_from_path`.

Also, `find_by_slug` and `find_root_by_slug` will have to be rewritten to search by `path` as well -- adding `/wiki/` back in, or potentially searching with a SQL wildcard like `%` in front of the `slug_from_path` result. I'm happy to offer ideas/help on that too, and it could become a separate issue for simplicity.

### PR DESCRIPTION
**Pull request by 500swapnil**
Make sure these boxes are checked before your pull request is ready to be reviewed and merged. Thanks!

* [x] all tests pass -- `rake test:all`
* [x] code is in uniquely-named feature branch, and has been rebased on top of latest master (especially if you've been asked to make additional changes)
* [x] pull request is descriptively named with #number reference back to original issue
* [x] if possible, multiple commits squashed if they're smaller changes
* [x] reviewed/confirmed/tested by another contributor or maintainer
* [x] `schema.rb.example` has been updated if any database migrations were added

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/wiki/contributing-to-public-lab-software

We have a loose schedule of reviewing and pulling in changes every Tuesday and Friday, and publishing changes on Fridays. Please alert developers on plots-dev@googlegroups.com when your request is ready or if you need assistance.

Thanks!

### CODE REVIEW

**Comment by jywarren** [comment]
Aha - i see this is using some ERB templating code in the test:

```
Failure: Expected response to be a redirect to <"http://test.host/wiki/revisions/<%= @node.slug %>?_=1481746385"> but was a redirect to <"http://test.host/wiki/revisions/<%= @node.slug_from_path %>?_=1481746385">.
```

For that we ought to use normal Ruby string substitution, like: `"/wiki/revisions/" + @node.slug` and `"/wiki/revisions/" + @node.slug_from_path` -- so both in the test and also in the redirect itself:

https://github.com/publiclab/plots2/pull/1101/files#diff-6d0eec853a873b030e87f5a21aecdf63L150

and

https://github.com/publiclab/plots2/blob/master/test/functional/admin_controller_test.rb#L324

It looks like that code was already wrong, actually! It was not running the ERB replacements, just spitting out the whole thing as a string, see? So it's great that we caught that in our revisions here.

**Comment by 500swapnil** [comment]
Yeah that is great! I shall make the changes now!

**Comment by 500swapnil** [comment]
Done! So the @node.slug_from_path function cannot be accessed from a Ruby string and needs to be added separately. That was the problem?

**Comment by jywarren** [comment]
Ah, looks like perhaps it should be `@revision.node.slug_from_path` since in that action we haven't defined `@node`. Fingers crossed that that'll get tests to pass!

**Comment by 500swapnil** [comment]
I defined the function slug_from_path in drupal_node.rb. So shouldnt it be @DrupalNode.slug_from_path?

**Comment by jywarren** [comment]
DrupalNode is the classname -- so you can use it to make a new node like
DrupalNode.new. But once you make one, like:

node = DrupalNode.new

then you refer to it by the variable you've stored the instance in. In this
case, we didn't make a `node` or `@node` variable, but we can access the
node we want because `@revision` has a `.node` method which fetches it.

**Comment by 500swapnil** [comment]
Okay

**Comment by jywarren** [comment]
Ah - sorry, looks like we need `@revision.drupal_node` not `.node` -- apologies!

**Comment by 500swapnil** [comment]
No problem. I had that doubt

**Comment by 500swapnil** [comment]
Same error

**Comment by jywarren** [comment]
Checking now! Also, you can always ping `@publiclab/reviewers` for help if i'm not available -- no guarantee, but sometimes if I'm busy with other work and you really want to make progress, it's worth it. Thanks!

File: test/functional/admin_controller_test.rb [hunk 1]
```diff
@@ -321,7 +321,7 @@ def teardown
     assert_equal 0, revision.status
     assert_equal 0, revision.author.status
     assert_not_equal node(:spam_targeted_page).latest.vid, revision.vid
-    assert_redirected_to "/wiki/revisions/<%= @node.slug %>" + '?_=' + Time.now.to_i.to_s
+    assert_redirected_to "/wiki/revisions/" + @revision.drupal_node.slug_from_path + '?_=' + Time.now.to_i.to_s
```
Comment (jywarren) [inline test/functional/admin_controller_test.rb]: OK -- while in the controller, we used an instance variable (starting with `@`) in the test, we are just using a local variable, so no `@` required. Check each changed line to ensure we're using what's available there -- either `@node` or `@revision` or `node` or `revision` -- and match the usage in that context.

File: app/views/wiki/show.html.erb [hunk 2]
```diff
@@ -77,7 +76,7 @@
       <% if @node.power_tags('tabbed').include?("notes") %>
         <div class="tab-pane" id="tab-notes">
           <%= render :partial => "notes/notes" %>
-          <p><a href="/tag/<%= @node.slug %>"><%= raw t('wiki.show.more_research', :tag => @node.slug) %> &raquo;</a></p>
+          <p><a href="/tag/<%= @revision.drupal_node.slug_from_path %>"><%= raw t('wiki.show.more_research', :tag => @revision.drupal_node.slug_from_path) %> &raquo;</a></p>
```
Comment (jywarren) [inline app/views/wiki/show.html.erb]: Yes -- see how here, we can indeed use `@node` -- it was only in some contexts where `@revision` was required, not all. Thanks!

Comment (500swapnil) [inline app/views/wiki/show.html.erb]: I went through the errors. @node is fine for all except admin_controller.rb and admin_controller_test.rb.
What should I use in admin_controller.rb and admin_controller_test.rb??

Comment (500swapnil) [inline app/views/wiki/show.html.erb]: Is there any way of testing a part of code and not the whole code locally? Some command instead of rake test:all .

Comment (jywarren) [inline app/views/wiki/show.html.erb]: On this line in the admin controller test, `revision.drupal_node`:

https://github.com/publiclab/plots2/blob/master/test/functional/admin_controller_test.rb#L324

And on this line, `@revision.drupal_node`:

https://github.com/publiclab/plots2/blob/master/app/controllers/admin_controller.rb#L150

Yes, you can run just functional tests with `rake test:functionals`, units with `rake test:units`, integration with `rake test:integration`, and you can even run just one test file with:

`ruby -I test test/functional/notes_controller_test.rb` (for example)

I know it's time consuming to run them all!

Comment (500swapnil) [inline app/views/wiki/show.html.erb]: Okay I will work on it now!

**Comment by 500swapnil** [comment]
Fine now?

**Comment by jywarren** [comment]
I'll take a look but I have to head out now -- later tonight I hope! Thanks!

**Comment by 500swapnil** [comment]
Okay I will have a look at other issues till then!

File: Gemfile.lock [hunk 3]
```diff
@@ -339,3 +339,9 @@ DEPENDENCIES
   uglifier (>= 1.0.3)
   will_paginate (>= 3.0.6)
   will_paginate-bootstrap (>= 1.0.1)
+
+RUBY VERSION
```
Comment (jywarren) [inline Gemfile.lock]: I'd like to not include Gemfile or Gemfile.lock changes if not essential, but I'll leave these in this time. Thanks!

**Comment by jywarren** [comment]
Looks great! merging now. Thanks for your careful work on this one!

**Comment by 500swapnil** [comment]
Thank you! Any idea on which issue I should try next?

**Comment by jywarren** [comment]
Yes -- there are some sub-issues to this one that'd be great to have some assistance on! https://github.com/publiclab/plots2/issues/397

For example, the checks on the `edit` and `update` actions, and the corresponding tests!

**Pull request merged**